File: examples/python/ex10_multi_model/main.py
"""10 Multi-Model — parallel comparison, cost routing, load balancing,
fallback chain, ensemble with judge.

Requires ``OPENAI_API_KEY`` and ``llm:gpt-4o`` + ``llm:gpt-4o-mini`` in
``resources.yaml``. Run from this directory:

    uv sync
    cp .env.example .env
    uv run python main.py
"""

from __future__ import annotations

import asyncio

import operonx
from operonx.core import END, PARENT, START, Operon, graph, op
from operonx.core.ops.flow.branch_op import if_
from operonx.providers import LLMOp


@op
def is_simple(classification: str):
    return {"is_simple": "SIMPLE" in classification.upper()}


@op
def compare(a, b):
    return {"gpt4o": a, "gpt4o_mini": b, "same_length": abs(len(a) - len(b)) < 50}


@op
def select(choice, a1, a2):
    return {
        "answer": a1 if "1" in choice else a2,
        "chosen": "gpt-4o" if "1" in choice else "gpt-4o-mini",
    }


@graph
def parallel_comparison(query):
    """Fan out to both gpt-4o and gpt-4o-mini in parallel → compare lengths."""
    gpt4o = LLMOp.of(
        resource="gpt-4o",
        prompt={"system": "Answer in one sentence.", "user": "{query}"},
        query=query,
    )
    gpt4o_mini = LLMOp.of(
        resource="gpt-4o-mini",
        prompt={"system": "Answer in one sentence.", "user": "{query}"},
        query=query,
    )
    cmp = compare(a=gpt4o["content"], b=gpt4o_mini["content"])
    START >> [gpt4o, gpt4o_mini] >> cmp >> END


@graph
def cost_routing(query):
    """Classify SIMPLE/COMPLEX, route cheap/expensive accordingly."""
    classifier = LLMOp.of(
        resource="gpt-4o-mini",
        prompt={
            "system": "Classify if this query is SIMPLE or COMPLEX. Reply with just one word.",
            "user": "{query}",
        },
        query=query,
    )
    check = is_simple(classification=classifier["content"])
    router = if_(check["is_simple"], "simple_llm").else_("complex_llm")

    simple_llm = LLMOp.of(
        resource="gpt-4o-mini",
        prompt={"system": "Be concise.", "user": "{query}"},
        query=query,
    )
    complex_llm = LLMOp.of(
        resource="gpt-4o",
        prompt={"system": "Think step by step.", "user": "{query}"},
        query=query,
    )

    simple_llm["content"] >> PARENT["answer"]
    complex_llm["content"] >> PARENT["answer"]

    START >> classifier >> check >> router
    router >> [simple_llm, complex_llm] >> END


@graph
def load_balanced(query):
    """Weighted model selection — 70% gpt-4o-mini, 30% gpt-4o."""
    llm = LLMOp.of(
        resource=["gpt-4o-mini", "gpt-4o"],
        ratios=[0.7, 0.3],
        prompt={"system": "Answer briefly.", "user": "{query}"},
        query=query,
    )
    START >> llm >> END


@graph
def fallback_chain(query):
    """gpt-4o with gpt-4o-mini as fallback."""
    llm = LLMOp.of(
        resource="gpt-4o",
        fallback=["gpt-4o-mini"],
        prompt={"system": "Answer briefly.", "user": "{query}"},
        query=query,
    )
    START >> llm >> END


@graph
def ensemble(query):
    """Two models in parallel; a judge LLM picks the better answer."""
    model_a = LLMOp.of(
        resource="gpt-4o",
        prompt={"system": "Answer the question accurately in 1-2 sentences.", "user": "{query}"},
        query=query,
    )
    model_b = LLMOp.of(
        resource="gpt-4o-mini",
        prompt={"system": "Answer the question accurately in 1-2 sentences.", "user": "{query}"},
        query=query,
    )

    judge = LLMOp.of(
        resource="gpt-4o-mini",
        prompt={
            "system": "Given a question and two answers, reply with just '1' or '2' for the better answer.",
            "user": "Question: {query}\n\nAnswer 1: {a1}\n\nAnswer 2: {a2}",
        },
        query=query,
        a1=model_a["content"],
        a2=model_b["content"],
    )
    sel = select(choice=judge["content"], a1=model_a["content"], a2=model_b["content"])
    START >> [model_a, model_b] >> judge >> sel >> END


async def main() -> None:
    operonx.bootstrap()

    runs = [
        (
            "parallel",
            parallel_comparison(query=PARENT["query"]),
            {"query": "What is machine learning?"},
        ),
        (
            "routing",
            cost_routing(query=PARENT["query"]),
            {"query": "Explain supervised vs unsupervised learning with examples."},
        ),
        ("load_balanced", load_balanced(query=PARENT["query"]), {"query": "Say hello #1"}),
        ("fallback", fallback_chain(query=PARENT["query"]), {"query": "What is Python?"}),
        (
            "ensemble",
            ensemble(query=PARENT["query"]),
            {"query": "What causes the seasons on Earth?"},
        ),
    ]
    for label, g, inputs in runs:
        try:
            result = await Operon(g).run(inputs=inputs)
            content = {k: v for k, v in result.items() if k != "$state"}
            print(f"[{label}] {content}")
        except Exception as e:
            print(f"[{label}] error: {e!r}")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/rust/ex10_multi_model/src/main.rs
//! 10 Multi-Model — Rust-side demo.
//!
//! Mirrors `examples/python/ex10_multi_model/main.py`. Only plain
//! `#[op]`s are declared; `PromptOp` + `LLMOp` are runtime-built.
//!
//! `routing` uses `if_()` which is stubbed in the Rust scheduler today;
//! it's excluded from runtime here.

use operonx::{op, BootstrapOpts, Operon};
use serde_json::Value;

#[op(name = "is_simple")]
fn is_simple(classification: String) -> Value {
    serde_json::json!({ "is_simple": classification.to_uppercase().contains("SIMPLE") })
}

#[op(name = "compare")]
fn compare(a: String, b: String) -> Value {
    let la = a.chars().count() as i64;
    let lb = b.chars().count() as i64;
    serde_json::json!({
        "gpt4o": a,
        "gpt4o_mini": b,
        "same_length": (la - lb).abs() < 50,
    })
}

#[op(name = "select")]
fn select(choice: String, a1: String, a2: String) -> Value {
    let pick_one = choice.contains('1');
    serde_json::json!({
        "answer": if pick_one { a1 } else { a2 },
        "chosen": if pick_one { "gpt-4o" } else { "gpt-4o-mini" },
    })
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let here = std::path::Path::new(env!("CARGO_MANIFEST_DIR"));

    let opts = BootstrapOpts::new().resources(here.join("resources.yaml"));
    operonx::bootstrap(opts);

    let graph_bundle: Value =
        serde_json::from_slice(&std::fs::read(here.join("graph.json"))?)?;
    let inputs_bundle: Value =
        serde_json::from_slice(&std::fs::read(here.join("inputs.json"))?)?;

    // `routing` skipped — needs if_() which is stubbed in Rust scheduler.
    for name in ["parallel", "load_balanced", "fallback", "ensemble"] {
        let graph_v = graph_bundle
            .get(name)
            .ok_or_else(|| format!("graph.json missing `{name}` entry"))?;
        let graph_json = serde_json::to_string(graph_v)?;

        let inputs_obj = inputs_bundle
            .get(name)
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default();

        let engine = Operon::builder(&graph_json).auto_register().build()?;
        match engine.run_json(inputs_obj, None, None, None) {
            Ok(r) => println!("[{name}] {r}"),
            Err(e) => println!("[{name}] error: {e}"),
        }
    }

    Ok(())
}
